l10n_it_fatturapa_out/wizard/efattura.py

```diff
@@ -255,9 +258,28 @@ def get_importo(line):
                 return False
             return line.price_unit * line.discount / 100
 
-        def get_default_note_tax(record):
-            all_taxes = get_all_taxes(record)
-            return all_taxes[0]["AliquotaIVA"] if all_taxes else "0.00"
+        def get_tax_ids(line):
+            """Get the tax for an invoice line.
+
+            Note: SdI expects only one tax per line. Also, it expects to find a tax line
+            for each line mentioned in the invoice lines, but Odoo generate no tax line
+            if the amount is 0. This is handled in get_all_taxes().
+
+            Common best practice for note/section lines is to reference an existing tax
+            line, instead of creating a new zero tax summary in the XML. The price for
+            note/section lines is zero anyway.  Therefore the original tax is ignored as
+            it defaults to the company default tax which may or may not be non zero in
+            the invoice."""
+
+            if line.display_type in ("line_section", "line_note"):
+                # find a non-zero tax, if possible
+                tax_lines = line.move_id.line_ids.filtered(
+                    lambda line: line.tax_ids
+                    and not (float_is_zero(line.price_total, precision_digits=2))
```

Comment:
> > Per controllare se un prezzo è 0, meglio usare odoo.tools.float_utils.float_is_zero per evitare problemi di precisione.
> 
> Fatto.

È meglio prendere la precisione del campo che stai controllando invece di scolpire `2`, in questo caso sarebbe la precisione di `price_total`.

> > Se ho ben capito qui stai cercando una riga che ha ammontare diverso da 0 perché sai che poi sarà anche inclusa nel riepilogo; non sarebbe più chiaro se cercassi direttamente un'imposta qualsiasi del riepilogo?
> 
> Potrei ma questa operazione, get_tax_ids(), avviene riga per riga prima di arrivare al riepilogo, di fatto in un template qweb diverso.

Per avere il riepilogo non è sufficiente chiamare `get_all_taxes`?

> > Inoltre vedo che hai aggiunto una parte in get_all_taxes proprio per aggiungere nel riepilogo le righe che hanno import 0, come mai allora qui devi cercare per forza le righe che hanno importo 0? Dalla modifca a get_all_taxes pare che anche le righe con import 0 finiranno in DatiRiepilogo.
>
> In get_all_taxes() vengono aggiunte al riepilogo le aliquote con imposta a zero. Sono relative a righe di fattura con imponibile diverso da zero ma zero imposta (totale). Quella di aggiungere una stanza nella get_all_taxes() è una strada già provata. Pare che allo SdI non piaccia una imposta a zero relativa solo a righe con zero imponibile (come le note).

Capito, grazie del chiarimento 👍🏻
Mi aveva confuso il `0 valued` nel commento del codice, potresti distinguere anche lì quando parli di 0 imponibile/ 0 aliquota/ 0 totale? Così sarà chiaro anche ai posteri

Reply:
> > > Se ho ben capito qui stai cercando una riga che ha ammontare diverso da 0 perché sai che poi sarà anche inclusa nel riepilogo; non sarebbe più chiaro se cercassi direttamente un'imposta qualsiasi del riepilogo?
> > 
> > Potrei ma questa operazione, get_tax_ids(), avviene riga per riga prima di arrivare al riepilogo, di fatto in un template qweb diverso.
>
> Per avere il riepilogo non è sufficiente chiamare `get_all_taxes`?

Parlandone stamattina in chiamata abbiamo chiarito che non c'è più il riepilogo come in v12, il _riepilogo_ di v14 è un sottoinsieme delle righe fattura e già stai cercando lì.
Non è necessario chiamare `get_all_taxes` perché aggiunge delle righe a 0 totale mentre qui ne vogliamo una che abbia un totale diverso da 0.

Reply:
Ho comunque eliminato il commento nel codice, quello della docstring del metodo dovrebbe già spiegare i senso della cosa.